### ISSUE TITLE
[feature] User configurable shell for user steps

### ISSUE THREAD

**Issue by minz1027** [screwdriver-cd/screwdriver#1229]
### Context
Currently we execute all cmds with `/bin/sh`. It's giving a lot of troubles to users when we force them to write sh cmds. If we can provide a way for them to configure the shell they want for user steps. That will be helpful.

### Proposed Solution
https://github.com/screwdriver-cd/launcher takes an new env, e.g. `USER_BIN_SHELL`. And use that to set the shebang for user steps. Need to do a check to split setup cmds and user cmds.

https://github.com/screwdriver-cd/launcher/blob/eb15cc9b533a0c901677593031d534615c763bd6/executor/executor.go#L241

**Comment by drewfish** [comment]
BTW, I was having trouble with a step in `screwdriver.yaml` which called a `foo.sh` which had `#!/bin/bash` and in that script I did a `sd-cmd`. Not sure if the issue was that `sd-cmd` didn't like bash or if it didn't like being called from a subprocess (or some other reason of course).

**Comment by d2lam** [comment]
I started working on this. However, it's more complicated than the original proposal. 

If start the shell as `sh`: ```c := exec.Command("/bin/sh")```
And for user steps we use `#!/bin/bash`
The step will still use `sh` because the launcher `source` the file, not `execute` the file. So it will still run in the same shell.

Example:
```
package main

import (
    "fmt"
    "gopkg.in/kr/pty.v1"
    "os/exec"
    "io"
    "os"
)

func main() {
    c := exec.Command("/bin/sh")
    f, err := pty.Start(c)
    if err != nil {
        fmt.Errorf("Cannot start shell: %v", err)
    }
    f.Write([]byte(". /tmp/step.sh\n"))
    f.Write([]byte{4})
    io.Copy(os.Stdout, f)
```

### Proposal
The reason that we source the `step.sh` file is to preserve the ssh-agent and environment variables. Currently, all SD setup steps & user build steps run in the same shell. We can try splitting them up like this:

- All SD set up steps run in the same `sh` shell
- All User steps run in the same shell that user configured (ex: `bash`). In the beginning, export the env from SD setup. 
- Each user teardown step runs in a separate user configured shell (ex: `bash`)
- Each SD teardown step runs in a separate `sh`

**Comment by drewfish** [comment]
If all the steps of the job run in the same shell, what about restating the feature request as "user configurable shell for each job". That hopefully is as easy as changing `c := exec.Command("/bin/sh")` to `c := exec.Command("/bin/bash")`.

Or perhaps just change the shell to bash (always, for everyone).

**Comment by d2lam** [comment]
@drewfish The reasons why we want the`sd-setup` and `sd-teardown` to run in `sh` is because they were written for `sh`. Some commands will work in `sh` and not `bash`, and vice versa. We can change them to use `bash` (this is easy by passing in a flag), but still probably shouldn't be the same as the shell that users configured, in case our bookends are not compatible. 

So our design is to have `sd-setup` and `sd-teardown` run in `sh` (or `bash`) --- hardcoded to what the cluster configured. And `user steps` and `user teardown` run in whatever the user configured using an environment variable.

**Comment by jithin1987** [comment]
Commands written for `sh` should run in `bash`, it's the otherway which won't work. I would do @drewfish suggestion, and use user configured shell for all steps.  Starting with how `launcher` starts the whole build process.

**Comment by d2lam** [comment]
Hmmm, maybe. I'm not sure since we cannot test all commands. For example, recently I ran this command (for launcher test) that works in `sh` but not in `zsh`:
```
(. /tmp/testTeardownEnv_export || echo 'Failed to get ENV' ) && echo bye 
-zsh(82903,0x7fff98ad2380) malloc: *** error for object 0x7fdcfe41f1a0: pointer being freed was not allocated
*** set a breakpoint in malloc_error_break to debug
```

I made a PR for the simpler implementation that Drew suggested: https://github.com/screwdriver-cd/launcher/pull/207

The more complicated implementation is in this branch: https://github.com/screwdriver-cd/launcher/pull/204. If we find out there are some complications, we can come back to this.
